src/hipscat_import/soap/resume_plan.py

```diff
@@ -0,0 +1,155 @@
+            - source pixel
+            - object pixels (healpix pixel with both order and pixel, for aligning and
+              neighboring object pixels)
+            - source key (string of source order+pixel)
+
+        """
+        counted_keys = set(self.read_log_keys(self.COUNTING_STAGE))
+        self.count_keys = [
+            (hp_pixel, object_pixels, f"{hp_pixel.order}_{hp_pixel.pixel}")
+            for hp_pixel, object_pixels in source_pixel_map.items()
+            if f"{hp_pixel.order}_{hp_pixel.pixel}" not in counted_keys
+        ]
+
+
+def source_to_object_map(args):
+    """Build a map of (source order/pixel) to the (object order/pixel)
+    that are aligned, as well as neighboring object pixels.
+    """
+    object_catalog = Catalog.read_from_hipscat(args.object_catalog_dir)
+    source_catalog = Catalog.read_from_hipscat(args.source_catalog_dir)
+
+    ## Direct aligment from source to object
+    ###############################################
+    grouped_alignment = align_trees(
+        object_catalog.pixel_tree, source_catalog.pixel_tree, "outer"
+    ).pixel_mapping.groupby(
+        [
+            PixelAlignment.JOIN_ORDER_COLUMN_NAME,
+            PixelAlignment.JOIN_PIXEL_COLUMN_NAME,
+        ],
+        group_keys=False,
+    )
+
+    ## Lots of cute comprehension is happening here.
+    ## create tuple of (source order/pixel) and [array of tuples of (object order/pixel)]
+    source_to_object = [
+        (
+            HealpixPixel(int(source_name[0]), int(source_name[1])),
+            [
+                HealpixPixel(int(object_elem[0]), int(object_elem[1]))
+                for object_elem in object_group.dropna().to_numpy().T[:2].T
+            ],
+        )
+        for source_name, object_group in grouped_alignment
+    ]
+    ## Treat the array of tuples as a dictionary.
+    source_to_object = dict(source_to_object)
+
+    ## Object neighbors for source
+    ###############################################
+    max_order = max(
+        object_catalog.partition_info.get_highest_order(),
+        source_catalog.partition_info.get_highest_order(),
+    )
+
+    object_order_map = np.full(hp.order2npix(max_order), -1)
+
+    for pixel in object_catalog.partition_info.get_healpix_pixels():
+        explosion_factor = 4 ** (max_order - pixel.order)
+        exploded_pixels = [
+            *range(
+                pixel.pixel * explosion_factor,
+                (pixel.pixel + 1) * explosion_factor,
+            )
+        ]
+        object_order_map[exploded_pixels] = pixel.order
+
+    for source, objects in source_to_object.items():
+        # get all neighboring pixels
+        nside = hp.order2nside(source.order)
+        neighbors = hp.get_all_neighbours(nside, source.pixel, nest=True)
+
+        ## get rid of -1s and normalize to max order
+        explosion_factor = 4 ** (max_order - source.order)
+        neighbors = [neighbor * explosion_factor for neighbor in neighbors if neighbor != -1]
```

Comment:
I think this isn't exactly getting the neighboring pixels after it's exploded, since there are `explosion_factor` number of pixels at a higher order, and this just computes the first of those, and the first pixel is always the top one. 

I don't think this matters after it has been desploded unless the object order is higher than the source order which probably shouldn't happen

Reply:
Having said that, I think Max has code in hipscat that gets the correct neighboring pixels as part of the margin cache